### PR TITLE
Fix MergeOp when using reduce pooling ops

### PR DESCRIPTION
**Pull request by AvinashBukkittu**
This PR

- Fixes a bug during merging of convolutional layers. Specifically, when reduce pooling operations are used, the feature dim is removed and hence merging happens along channel dim
- Adds a test case in `Conv1DNetworkTest` to verify the above behavior.
- Tests the output of `MergeLayer` for all modes

This fixes https://github.com/asyml/texar-pytorch/issues/44

### CODE REVIEW

File: texar/core/layers.py [hunk 1]
```diff
@@ -664,6 +664,14 @@ def forward(self, input: torch.Tensor) -> torch.Tensor:  # type: ignore
                 layer_output = layer(input)
                 layer_outputs.append(layer_output)
 
+        # the merge dimension cannot be determined until we get the output from
+        # individual layers.
+        # In case of reduce pooling operations, feature dim is removed and
+        # channel dim is merged.
+        # In non-reduce pooling operations, feature dim is merged.
+        if self._dim is None:
+            self._dim = layer_outputs[0].dim() - 1
```
Comment (huzecong) [inline texar/core/layers.py]: Prefer not to change `self._dim`. It might be possible for the user to pass in tensors of different ranks.

File: texar/core/layers.py [hunk 2]
```diff
@@ -664,6 +664,14 @@ def forward(self, input: torch.Tensor) -> torch.Tensor:  # type: ignore
                 layer_output = layer(input)
```
Comment (huzecong) [inline texar/core/layers.py]: (This comment is for line 660 but somehow I can't add a comment there.)

If `self._layers` is `None`, is it valid for `input` to be a tensor? To me it seems more reasonable if `input` is a list of tensors. This way the behaviors between cases w/ and w/o `self._layers` are consistent. Also, `torch.cat` requires the argument to be a list of tensors.

File: texar/core/layers.py [hunk 3]
```diff
@@ -664,6 +664,14 @@ def forward(self, input: torch.Tensor) -> torch.Tensor:  # type: ignore
...
+        # individual layers.
+        # In case of reduce pooling operations, feature dim is removed and
+        # channel dim is merged.
+        # In non-reduce pooling operations, feature dim is merged.
+        if self._dim is None:
+            self._dim = layer_outputs[0].dim() - 1
+
         if self._mode == 'concat':
             outputs = torch.cat(tensors=layer_outputs, dim=self._dim)
         elif self._mode == 'elemwise_sum':
```
Comment (huzecong) [inline texar/core/layers.py]: (This comment is for lines 685-708.)

The logic here is almost the same, maybe we can unify them by e.g. creating a class-level dictionary mapping modes to torch functions and just looking them up.

File: texar/core/layers.py (hunk 1 above)
Comment (AvinashBukkittu) [reply texar/core/layers.py]: I considered that possibility. `dim` arguments to torch functions in `forward` depends on the size of the input tensor. I am thinking to remove `self._dim` and make it function variable. In that case, we should be not be taking `dim` argument in the constructor. Also, `dim` is not being passed with any value during construction of `MergeLayer`. I think we can safely remove this. However, if you consider `texar-TF` interface compatibility, we may have to keep this.

File: texar/core/layers.py (hunk 2 above)
Comment (AvinashBukkittu) [reply texar/core/layers.py]: `input` is the input to conv network or output of previous layers of the conv network, so as far as I know `input` is a `torch.Tensor`. The `if self._layers is None` handles a corner case where we allow `MergeLayer` to have no layers, in which case, we just return the input tensor. What we can do is, if the `if` condition holds true, we can set `layer_outputs = [inputs]`, in that way, `layer_outputs` is always a list.

File: texar/core/layers.py (hunk 3 above)
Comment (AvinashBukkittu) [reply texar/core/layers.py]: True, I agree with that. Removes a lot of repetition. This was copied with little to no changes from `texar-TF`.

File: texar/core/layers.py (hunk 1 above)
Comment (huzecong) [reply texar/core/layers.py]: We don't really need to move the `dim` argument to `forward`. What I meant was:
```python
dim = self._dim is self._dim is not None else -1
```
and use `dim` in the merging ops.

File: texar/core/layers.py (hunk 2 above)
Comment (huzecong) [reply texar/core/layers.py]: So when `self._layers` is `None`, `MergeLayer` acts as identity? I don't really think we need to handle this; we just don't construct the `MergeLayer` in that case, which saves an unnecessary function call during `forward`.

If we do need to support that case, then we should just return the `input` instead of executing the rest of the logic. And we should have a test case for that.

File: texar/core/layers.py [hunk 4]
```diff
@@ -647,6 +647,24 @@ def __init__(self, layers: Optional[List[nn.Module]] = None,
...
+            "or": torch.any,
+            "logsumexp": torch.logsumexp
+        }
+
+        def merge(tensors, dim):
+            _concat = torch.cat(tensors=tensors, dim=dim)
+            outputs = self._functions[self._mode](_concat, dim=dim)
+            return outputs
+
+        self._merge_op = merge
```
Comment (huzecong) [inline texar/core/layers.py]: I don't think it's necessary to store a function as attribute. The logic is simple and can be expanded in `forward` directly.

File: texar/core/layers.py (hunk 1 above)
Comment (AvinashBukkittu) [reply texar/core/layers.py]: Done!

File: texar/core/layers.py (hunk 2 above)
Comment (AvinashBukkittu) [reply texar/core/layers.py]: Yes, it acts as an Identity operator. As discussed offline, we will provide this feature to maintain compatibility with `texar-TF`. I have added a test case to cover `self._layers = None` case.

File: texar/core/layers.py (hunk 4 above)
Comment (AvinashBukkittu) [reply texar/core/layers.py]: Done!
